### PR TITLE
refactor(data)!: unify multimodal dataset builders

### PR DESCRIPTION
**Pull request by yaoyu-33**
## What does this PR do?

Completes the VLM preloaded-data cleanup phase of NVIDIA-NeMo/Megatron-Bridge#4664 on top of NVIDIA-NeMo/Megatron-Bridge#4685 and extends the same Config + Builder boundary to every supported multimodal data path.

Supported paths after this PR:

* Direct Hugging Face / conversation SFT through `DirectHFSFTDatasetConfig` + `DirectHFSFTDatasetBuilder`, including local JSON/JSONL through the Hugging Face datasets `json` loader.
* Sharded multimodal WebDataset through `EnergonDatasetConfig` + `EnergonDatasetBuilder`.
* Synthetic VLM recipe data through `MockVLMSFTDatasetConfig` + `MockVLMSFTDatasetBuilder`.

Runtime processors, tokenizers, task encoders, collators, and datasets are constructed by builders; serialized configs contain only declarative values.

## Scope

* Remove `PreloadedVLMConversationProvider`, `vlm-preloaded`, and the obsolete local-conversation source path.
* Add typed, CLI-overridable Energon task-encoder configs for generic HF, Qwen-VL, and Nemotron Omni.
* Migrate Qwen-VL and Nemotron Omni Energon recipes away from eager processor/tokenizer/task-encoder construction.
* Remove `EnergonProvider` and `QwenVLEnergonProvider`; `--dataset vlm-energon` now requires a recipe with a complete model-specific `EnergonDatasetConfig`.
* Replace `MockVLMConversationProvider` in Qwen3-VL/Qwen3.5-VL recipes and functional helpers.
* Delete the entire `megatron.bridge.data.vlm_datasets` package.
* Move shared visual-key constants to `data.collators.visual` and replace the eager collator map with a lazy processor-type resolver in `data.collators.registry`.
* Move the Step3.7 Flickr8k pipeline to `models.stepfun.data.flickr8k`.
* Preserve chat-template/loss-mask behavior, processor-native media references on the Direct-HF path (without preloaded-schema adaptation), VLM/omni collators, pure-DP Energon worker sharding, CP/SP padding, deferred Qwen packing, collate-time packing for other models, and the absence of an Energon test split.
* Migrate the Qwen3-Omni JSONL recipe/launcher to the Direct-HF JSON path, add its model-owned Direct-HF collator, and keep the full-SFT default at 8 H100s (TP1/PP1/EP8).
* Split the Hugging Face tutorials into focused `hf-text-only` and `hf-multimodal` paths; start multimodal users from a hosted MedPix source before local JSON, Qwen3-VL, packing, and migration guidance; retain the separate Energon tutorial and real MedPix-to-Energon helper.

## Migration notes

| Removed API/path | Replacement |
| -- | -- |
| `PreloadedVLMConversationProvider`, `--dataset vlm-preloaded` | `DirectHFSFTDatasetConfig` with `HFDatasetSourceConfig(path_or_dataset="json", load_kwargs={"data_files": ...})`, or Energon |
| `EnergonProvider` | `EnergonDatasetConfig` + `EnergonDatasetBuilder` |
| `EnergonMultiModalDataModule(image_processor=..., global_batch_size=...)` | Remove those constructor arguments; processors are runtime task-encoder assets and global batch size remains in the training config |
| `QwenVLEnergonProvider` fields such as `dataset.max_num_images` | `dataset.task_encoder.max_num_images` on `QwenVLEnergonTaskEncoderConfig` |
| `MockVLMConversationProvider` | `MockVLMSFTDatasetConfig` + `MockVLMSFTDatasetBuilder` |
| `data.vlm_datasets.collate.COLLATE_FNS` | `data.collators.registry.resolve_model_collate`; collator implementations remain model-owned |
| `data.vlm_datasets.collate_utils.{THW_GRID_VISUAL_KEYS,PASSTHROUGH_VISUAL_KEYS}` | `data.collators.visual` |
| `data.vlm_datasets.step37_flickr8k` | `models.stepfun.data.flickr8k` |
| `qwen3_omni_30b_a3b_sft_preloaded_config` | `qwen3_omni_30b_a3b_sft_hf_json_config` |

No deprecated provider, selector alias, local VLM source, or compatibility `vlm_datasets` package is retained.

## Validation

Focused static/unit validation in the `megatron-bridge:latest` development image:

* Energon builder/recipes/config: **120 passed**.
* Energon datamodule CP/worker/split behavior: **25 passed**.
* Mock VLM recipes/config/packing: **39 passed**.
* Model collators, Direct SFT, HF task encoder, Step3.7 relocation, removed-module and docs checks: **118 passed**.
* Lazy collator import-cycle subprocess tests: **2 passed**.
* Combined Energon/Mock/Qwen/CP-SP/in-batch packing selection: **96 passed, 1 skipped**.
* Earlier Direct-HF/local JSON/Qwen3-Omni migration selection: **128 passed**.
* Qwen3-Omni topology/launcher/docs selection: **55 passed**.
* `bash -n examples/models/qwen/qwen3_omni/local_train_thinker_full.sh`: passed.
* Independent review-fix regression (ConfigContainer, Energon builder, Direct SFT, Qwen3-VL/Qwen3.5-VL recipes): **407 passed**.
* Independent round-2 combined architecture/packing/recipe/module/docs matrix: **702 passed, 1 CUDA-only skipped**.
* Processor-native local JSON documentation regression: **54 passed**.
* Final Energon `--seq_length` and documentation regression: **57 passed**.
* Round-3 final builder/config/docs spot-check: **315 passed**.
* Follow-up Direct-HF/local JSON, Qwen3-Omni collator/step/recipe, CLI/tutorial, and docs matrix: **167 passed**.
* Final Qwen3-Omni collator/step/recipe regression, including mixed-length right-padding behavior: **19 passed**.
* MedPix tutorial, Hub preset/adapter, Direct-HF builder, Energon builder, Hydra slice quoting, and docs regression: **97 passed**.
* Hugging Face text-only/multimodal tutorial split, renamed helpers, hosted MedPix entry point, and cross-doc links: **22 passed**.
* A real `Qwen3OmniMoeProcessor` built and collated local JSON with image and audio through `DirectHFSFTDatasetBuilder`, producing the expected text, visual, audio, label, loss-mask, and position tensors.
* `uv run --active --no-sync pre-commit run --all-files`: passed on the final commit.

Focused GPU validation with Qwen3-VL 8B LoRA on one H100 (the Qwen3-Omni 30B full-SFT recipe itself remains an 8-GPU EP8 recipe):

* Direct-HF local JSON completed a training iteration: loss **4.314859**, no NaN/skip.
* Energon with micro-batch size 2, validation, and checkpointing completed: train loss **4.662656**, validation loss **1.658658**, checkpoint saved.
* Energon deferred in-batch packing completed a training iteration: loss **4.682606**, no NaN/skip.
* Direct-HF `medpix` preset completed 3 train iterations plus validation and checkpoint save: final train loss **1.591787**, validation loss **1.927279**, no NaN/skip.
* MedPix Energon unpacked completed 3 train iterations plus validation and checkpoint save: final train loss **2.734495**, validation loss **2.089008**, no NaN/skip.
* The same MedPix Energon shards with deferred in-batch packing completed 3 train iterations plus validation and checkpoint save: final train loss **2.738407**, validation loss **2.090850**, no NaN/skip.
* The checked-in MedPix helper completed real Hub slicing, tar generation, Energon indexing, manifest, and dataset metadata generation.
* The documented HF-to-native Qwen3-VL checkpoint import followed by a quoted `train[:4]` Direct-HF MedPix smoke completed: loss **3.106370**, no NaN/skip.
* [W&B Direct-HF/Energon comparison report](<https://wandb.ai/yaoyu/megatron-bridge-pr4774-medpix/reports/Megatron-Bridge-PR-4774:-MedPix-Direct-HF-and-Energon-E2E--VmlldzoxNzQ1OTQ3NA==>): all three runs finished at step 3 with finite train/validation loss and peak allocated GPU memory of **32.562 GB**.

The host environment cannot import the pinned MCore with its local NVRX development build, so Python tests used the project development image with `PYTHONPATH` pointing at this checkout. GPU E2E validation ran on the requested internal cluster as listed above.

## Remaining Issue NVIDIA-NeMo/Megatron-Bridge#4664 phases (out of scope)

* Implement the prepared `.bin`/`.idx` SFT path that will replace packed-Parquet as the scalable prepared text-SFT workflow.
* Migrate the remaining legacy text-SFT compatibility path once that replacement is available and representative recipes are verified.

This PR does not modify Megatron-LM, CI workflows, or dependencies.

## Final EOS example validation

Focused validation used the requested 260522 container and current PR checkout:

* Energon 7.x preparation completed for the checked-in tiny image/chat helper, a real one-row Mantis parquet/image conversion, and real MedPix Hub slices. The shared preparation helper also passes with Energon 7.4.
* Qwen3.5-VL MegatronMIMO Direct-HF CORD completed one 2-GPU iteration at sequence length 1024: loss **12.53403**, no skipped/NaN iterations.
* Qwen3-VL 8B LoRA with the checked-in Energon launcher completed one 1-GPU packed iteration plus validation: train loss **4.348903**, validation loss **4.511334**, no skipped/NaN iterations.
* Qwen3-Omni 30B-A3B full SFT used the documented 8-GPU TP1/PP1/EP8 topology and local HF JSON. With the launcher’s default optimizer CPU offload it completed one iteration: loss **1.776158**, load-balancing loss **2.498360**, no skipped/NaN iterations; peak allocated GPU memory was **64.742 GB**.
* `uv run --active --no-sync pre-commit run --all-files`: passed.

Runtime validation also fixed three example-only issues found during execution: Qwen3.5 visual-token truncation now fails with a clear sequence-budget error, empty Mantis split shards are removed before indexing, and image-only Energon loading no longer initializes the optional PyAV decoder.

### 20-iteration convergence checks

Both checks used real Hugging Face pretrained weights and deterministic four-example local datasets; every optimizer step consumed all four examples. These are intentional overfit/pipeline-convergence checks, not held-out quality benchmarks.

* **Qwen3-0.6B full SFT**, 1 H100, sequence length 512, local Hugging Face `json` prompt/completion rows through `GPTSFTDatasetConfig`: **20/20 iterations**, loss **5.916000 -> 0.000011134**, first-five mean **1.907456**, last-five mean **0.000011480**, OLS slope **-0.128111**; no skipped or NaN iterations.
* **Qwen3-VL-8B-Instruct LoRA**, 1 H100, sequence length 1024, processor-native local image JSON through `DirectHFSFTDatasetConfig`: **20/20 iterations**, loss **4.599410 -> 0.000007408**, first-five mean **1.298043**, last-five mean **0.000010816**, OLS slope **-0.089362**; no skipped or NaN iterations.
* **Qwen3-VL-8B-Instruct LoRA with generic** `vlm_step`, 1 H100, sequence length 1024, processor-native local image JSON through `DirectHFSFTDatasetConfig`, collate-time in-batch packing, and 2 data workers: **20/20 iterations**, loss **4.588531 -> 0.000010925**; no skipped or NaN iterations, and the final checkpoint saved successfully (CW job `13718109`, exit `0:0`).
* **Qwen3-VL-8B-Instruct Energon with generic** `vlm_step`, 1 H100, 2 train workers plus 2 validation workers, unpacked baseline: train loss **4.314859**, validation losses **1.710499** and **1.790600**; no skipped or NaN iterations (CW job `13718349`, exit `0:0`).

### CODE REVIEW

File: src/megatron/bridge/data/builders/energon.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        raise ValueError(f"{field_name} must be a non-empty string.")
+
+
+@dataclass(kw_only=True)
+class HFEnergonTaskEncoderConfig:
+    """Serializable settings for the generic Hugging Face Energon task encoder."""
+
+    hf_processor_path: str
+    visual_keys: tuple[str, ...] = ("pixel_values",)
+    min_pixels: int | None = None
```
Comment (yaoyu-33) [inline src/megatron/bridge/data/builders/energon.py:45]: min_pixels / max pixels are visual keys?

Comment (yaoyu-33) [reply src/megatron/bridge/data/builders/energon.py:45]: No—these are separate task-encoder settings. `visual_keys` lists processor output tensors retained in `GenericVisualInputs` (for example, `pixel_values` or `image_grid_thw`). `min_pixels` and `max_pixels` are scalar preprocessing bounds passed to the processor/collator to control visual resolution and token cost. They live in the same nested task-encoder config because it owns model-specific processor/collator behavior. I clarified that distinction in the config docstrings and the new Energon tutorial, and added explicit forwarding assertions in `0c9b1cea5`.

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4774/commits/e3eeab4585f4163694ab63b002b2f4dfd94f6989

File: tutorials/data/hf-multimodal/README.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+from megatron.bridge.recipes.qwen_vl import qwen3_vl_8b_peft_config
+
+cfg = qwen3_vl_8b_peft_config()
+cfg.dataset.source = HFDatasetSourceConfig(dataset_name="medpix")
+cfg.dataset.do_validation = True
+cfg.dataset.do_test = False
+```
+
+The recipe already selects `DirectHFSFTDatasetConfig`, Qwen's processor, assistant-only loss, and deferred
+in-batch packing. Replace only the source for another compatible preset or hosted dataset. When hosted rows are
```
Comment (cuichenx) [inline tutorials/data/hf-multimodal/README.md]: mention, normally do not use deferred in-batch packing

Comment (cuichenx) [reply tutorials/data/hf-multimodal/README.md]: for qwen3 you don't have to use deferred in-batch packing right?

File: tutorials/data/hf-multimodal/README.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+```bash
+export WANDB_MODE=disabled
+export OUTPUT_DIR="$WORKSPACE/results/qwen3-vl-hf-local-smoke"
+
+uv run python -m torch.distributed.run --standalone --nproc_per_node=1 \
+  scripts/training/run_recipe.py \
+  --recipe qwen3_vl_8b_peft_config \
+  --dataset vlm-hf \
+  --step_func qwen3_vl_step \
```
Comment (cuichenx) [inline tutorials/data/hf-multimodal/README.md]: use vlm_step

File: tutorials/data/energon/README.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  checkpoint.save_interval=1 \
+  train.train_iters=1 \
+  train.global_batch_size=2 \
+  train.micro_batch_size=2 \
+  validation.eval_interval=1 \
+  validation.eval_iters=1 \
+  validation.eval_micro_batch_size=2 \
+  dataset.path="$ENERGON_PATH" \
+  dataset.micro_batch_size=2 \
+  dataset.num_workers=0 \
```
Comment (yaoyu-33) [inline tutorials/data/energon/README.md]: why  workers=0? should use a more realistic case.

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4774/commits/d72b0ee0fa8b9281f9c4fd2d271e9e98ccbda4cf

File: tutorials/data/hf-multimodal/README.md (hunk 3 above)
Comment (yaoyu-33) [reply tutorials/data/hf-multimodal/README.md]: Updated in b0ffcbd9f: the Direct-HF tutorial and overview commands now use `--step_func vlm_step`. I reran Qwen3-VL 8B LoRA with local image JSON, collate-time in-batch packing, and `vlm_step`: 20/20 iterations completed, loss 4.588531 -> 0.000010925, no skipped/NaN iterations (CW 13718109, exit 0:0).

File: tutorials/data/energon/README.md (hunk 4 above)
Comment (yaoyu-33) [reply tutorials/data/energon/README.md]: Updated in b0ffcbd9f: the Energon tutorial now uses 2 train workers and 2 validation workers, with zero workers documented only as a debugging setting. The matching `vlm_step` smoke completed train + validation with finite losses and no skipped/NaN iterations (CW 13718349, exit 0:0).

File: tutorials/data/hf-multimodal/README.md (hunk 2 above)
Comment (yaoyu-33) [reply tutorials/data/hf-multimodal/README.md]: Updated in b0ffcbd9f: the tutorial now recommends the normal collate-time packing path (`defer_in_batch_packing_to_step=False`) and explains that Qwen3-VL does not need deferred packing for this workflow. The generic-path GPU rerun completed 20/20 iterations with no skipped/NaN iterations.

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4774/commits/f3404cb92ef6584090d27fdf573f7c7bae8ccae9

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4774/commits/55bd6733395f6b683f1aef8e5bfdc56e01a34acd

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4774/commits/414bb6bf2cd1d998ed3b7aaf76bdd755bdda3d28
